### PR TITLE
Add adamw and lr decay

### PR DESCRIPTION
**Pull request by kobiso**
This commit adds AdamW optimizer and LR decay with 'ReduceLROnPlateau' by following the DALLE paper.
In DALLE paper,
image
image
Both AdamW optimizer and LR decay in the code follows the same parameters of the paper.
I found LR decay is helpful as shown in https://github.com/lucidrains/DALLE-pytorch/discussions/131, but AdamW  optimizer is not stable yet.
So, I set both AdamW optimizer and LR decay is optional.
@lucidrains , Please check the code :)

### CODE REVIEW

**Comment by lucidrains** [comment]
@kobiso this looks great! I think LR decay is a good idea :)

So there is one thing to watch out for when doing weight decay with Transformers, and that is the layernorm weights and the biases should not be included

This can be achieved using param groups - I'll see if I can find an example, unless if you get to it before I wake up ;)

**Comment by kobiso** [comment]
ohhhh maybe that could be the reason of this : https://github.com/lucidrains/DALLE-pytorch/discussions/139
I am new to Transformer and learn lots from you! thanks :)
I will check what you mentioned!

**Comment by kobiso** [comment]
So, I added param groups function to disregard the bias and LayerNorm weight in https://github.com/lucidrains/DALLE-pytorch/pull/138/commits/fd7307530d414658f594b3216dc5ccb1db9264f6.
It is applied only to AdamW because current Adam setting does not use weight decay.
When Adam start using it, we can simply apply `group_weight()` function in the same manner.
These are the parameter lists set by `weight_decay=0.`:

<details>
<summary>Parameter lists with `weight_decay=0.`</summary>

['transformer.layers.layers.0.0.norm.weight',
 'transformer.layers.layers.0.0.norm.bias',
 'transformer.layers.layers.0.0.fn.to_out.0.bias',
 'transformer.layers.layers.0.1.norm.weight',
 'transformer.layers.layers.0.1.norm.bias',
 'transformer.layers.layers.0.1.fn.net.0.bias',
 'transformer.layers.layers.0.1.fn.net.3.bias',
 'transformer.layers.layers.1.0.norm.weight',
 'transformer.layers.layers.1.0.norm.bias',
 'transformer.layers.layers.1.0.fn.to_out.0.bias',
 'transformer.layers.layers.1.1.norm.weight',
 'transformer.layers.layers.1.1.norm.bias',
 'transformer.layers.layers.1.1.fn.net.0.bias',
 'transformer.layers.layers.1.1.fn.net.3.bias',
 'transformer.layers.layers.2.0.norm.weight',
 'transformer.layers.layers.2.0.norm.bias',
 'transformer.layers.layers.2.0.fn.to_out.0.bias',
 'transformer.layers.layers.2.1.norm.weight',
 'transformer.layers.layers.2.1.norm.bias',
 'transformer.layers.layers.2.1.fn.net.0.bias',
 'transformer.layers.layers.2.1.fn.net.3.bias',
 'transformer.layers.layers.3.0.norm.weight',
 'transformer.layers.layers.3.0.norm.bias',
 'transformer.layers.layers.3.0.fn.to_out.0.bias',
 'transformer.layers.layers.3.1.norm.weight',
 'transformer.layers.layers.3.1.norm.bias',
 'transformer.layers.layers.3.1.fn.net.0.bias',
 'transformer.layers.layers.3.1.fn.net.3.bias',
 'transformer.layers.layers.4.0.norm.weight',
 'transformer.layers.layers.4.0.norm.bias',
 'transformer.layers.layers.4.0.fn.to_out.0.bias',
 'transformer.layers.layers.4.1.norm.weight',
 'transformer.layers.layers.4.1.norm.bias',
 'transformer.layers.layers.4.1.fn.net.0.bias',
 'transformer.layers.layers.4.1.fn.net.3.bias',
 'transformer.layers.layers.5.0.norm.weight',
 'transformer.layers.layers.5.0.norm.bias',
 'transformer.layers.layers.5.0.fn.to_out.0.bias',
 'transformer.layers.layers.5.1.norm.weight',
 'transformer.layers.layers.5.1.norm.bias',
 'transformer.layers.layers.5.1.fn.net.0.bias',
 'transformer.layers.layers.5.1.fn.net.3.bias',
 'transformer.layers.layers.6.0.norm.weight',
 'transformer.layers.layers.6.0.norm.bias',
 'transformer.layers.layers.6.0.fn.to_out.0.bias',
 'transformer.layers.layers.6.1.norm.weight',
 'transformer.layers.layers.6.1.norm.bias',
 'transformer.layers.layers.6.1.fn.net.0.bias',
 'transformer.layers.layers.6.1.fn.net.3.bias',
 'transformer.layers.layers.7.0.norm.weight',
 'transformer.layers.layers.7.0.norm.bias',
 'transformer.layers.layers.7.0.fn.to_out.0.bias',
 'transformer.layers.layers.7.1.norm.weight',
 'transformer.layers.layers.7.1.norm.bias',
 'transformer.layers.layers.7.1.fn.net.0.bias',
 'transformer.layers.layers.7.1.fn.net.3.bias']

</details>

Also, I am running an experiment with this setup now.
Please check if this is correct :)

**Comment by lucidrains** [comment]
💯
